components/merkledb/src/proto/mod.rs

```diff
@@ -17,44 +17,143 @@
 #![allow(bare_trait_objects)]
 
-pub use self::list_proof::*;
+use failure::Error;
+use protobuf::{well_known_types::Empty, RepeatedField};
+use std::borrow::Cow;
+
+use crate::{proof_map_index::ProofPath, BinaryKey, BinaryValue};
 use exonum_crypto::proto::*;
+use exonum_proto::ProtobufConvert;
+
+pub use self::{list_proof::*, map_proof::*};
 
 include!(concat!(env!("OUT_DIR"), "/protobuf_mod.rs"));
 
+impl<K, V> ProtobufConvert for crate::MapProof<K, V>
+where
+    K: BinaryKey + ToOwned<Owned = K>,
+    V: BinaryValue,
+{
+    type ProtoStruct = MapProof;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        let mut map_proof = MapProof::new();
+
+        let proof: Vec<MapProofEntry> = self
+            .proof_unchecked()
+            .iter()
+            .map(|(p, h)| {
+                let mut entry = MapProofEntry::new();
+                entry.set_hash(h.to_pb());
+                entry.set_proof_path(p.as_bytes().to_vec());
+                entry
+            })
+            .collect();
+
+        let entries: Vec<OptionalEntry> = self
+            .all_entries_unchecked()
+            .map(|(key, value)| {
+                let mut entry = OptionalEntry::new();
+                let mut buf = vec![0_u8; key.size()];
+                key.write(&mut buf);
+                entry.set_key(buf.to_vec());
+
+                match value {
+                    Some(value) => entry.set_value(value.to_bytes()),
+                    None => entry.set_no_value(Empty::new()),
+                }
+
+                entry
+            })
+            .collect();
+
+        map_proof.set_proof(RepeatedField::from_vec(proof));
+        map_proof.set_entries(RepeatedField::from_vec(entries));
+
+        map_proof
+    }
+
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, Error> {
+        let proof = pb
+            .get_proof()
+            .iter()
+            .map(|entry| {
+                Ok((
+                    ProofPath::read(entry.get_proof_path()),
+                    exonum_crypto::Hash::from_pb(entry.get_hash().clone())?,
+                ))
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        let entries = pb
+            .get_entries()
+            .iter()
+            .map(|entry| {
+                let key = K::read(entry.get_key());
+
+                let value = if entry.has_value() {
+                    Some(V::from_bytes(Cow::Borrowed(entry.get_value()))?)
+                } else {
+                    None
+                };
+
+                Ok((key, value))
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        let mut map_proof = crate::MapProof::new().add_proof_entries(proof);
+
+        for entry in entries {
+            map_proof = match entry.1 {
+                Some(value) => map_proof.add_entry(entry.0, value),
+                None => map_proof.add_missing(entry.0),
+            };
+        }
+
+        Ok(map_proof)
+    }
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
I'm confused by changes here. Does this replace tests for `ListProof` serialization with tests for `MapProof`? They should be besides each other, don't they?